signature/jws/plugin.go

```diff
@@ -0,0 +1,185 @@
+		return s.generateSignatureEnvelope(ctx, desc, opts)
+	}
+	return nil, fmt.Errorf("plugin does not have signing capabilities")
+}
+
+func (s *PluginSigner) describeKey(ctx context.Context, config map[string]string) (*plugin.DescribeKeyResponse, error) {
+	req := &plugin.DescribeKeyRequest{
+		ContractVersion: "1",
+		KeyID:           s.KeyID,
+		PluginConfig:    config,
+	}
+	out, err := s.Runner.Run(ctx, plugin.CommandDescribeKey, req)
+	if err != nil {
+		return nil, fmt.Errorf("describe-key command failed: %w", err)
+	}
+	resp, ok := out.(*plugin.DescribeKeyResponse)
+	if !ok {
+		return nil, fmt.Errorf("plugin runner returned incorrect describe-key response type '%T'", out)
+	}
+	return resp, nil
+}
+
+func (s *PluginSigner) generateSignature(ctx context.Context, desc signature.Descriptor, opts notation.SignOptions) ([]byte, error) {
+	config := s.mergeConfig(opts.PluginConfig)
+	// Get key info.
+	key, err := s.describeKey(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check keyID is honored.
+	if s.KeyID != key.KeyID {
+		return nil, fmt.Errorf("keyID mismatch")
+	}
+
+	// Get algorithm associated to key.
+	alg := key.KeySpec.SignatureAlgorithm()
+	if alg == "" {
+		return nil, fmt.Errorf("keySpec %q not supported: ", key.KeySpec)
+	}
+
+	// Generate payload to be signed.
+	payload := packPayload(desc, opts)
+	if err := payload.Valid(); err != nil {
+		return nil, err
+	}
+
+	// Generate signing string.
+	token := jwtToken(alg.JWS(), payload)
+	signing, err := token.SigningString()
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal signing payload: %v", err)
+	}
+
+	// Execute plugin sign command.
+	req := &plugin.GenerateSignatureRequest{
+		ContractVersion: "1",
+		KeyID:           s.KeyID,
+		KeySpec:         key.KeySpec,
+		Hash:            alg.Hash(),
+		Payload:         signing,
+		PluginConfig:    config,
+	}
+	out, err := s.Runner.Run(ctx, plugin.CommandGenerateSignature, req)
+	if err != nil {
+		return nil, fmt.Errorf("generate-signature command failed: %w", err)
+	}
+	resp, ok := out.(*plugin.GenerateSignatureResponse)
+	if !ok {
+		return nil, fmt.Errorf("plugin runner returned incorrect generate-signature response type '%T'", out)
+	}
+
+	// Check keyID is honored.
+	if s.KeyID != resp.KeyID {
+		return nil, fmt.Errorf("keyID mismatch")
+	}
+
+	// Check algorithm is supported.
+	jwsAlg := resp.SigningAlgorithm.JWS()
+	if jwsAlg == "" {
+		return nil, fmt.Errorf("signing algorithm %q not supported", resp.SigningAlgorithm)
+	}
+
+	// Check certificate chain is not empty.
+	if len(resp.CertificateChain) == 0 {
+		return nil, errors.New("empty certificate chain")
+	}
+
+	certs, err := parseCertChain(resp.CertificateChain)
+	if err != nil {
+		return nil, err
+	}
+
+	// Verify the hash of the request payload against the response signature
+	// using the public key of the signing certificate.
+	signed64Url := base64.RawURLEncoding.EncodeToString(resp.Signature)
```

Comment:
Bit confused here, seems we are converting base64 encoded `resp.Signature`, the base64 encoded raw signature generated by the plugin, to base64URL encoded string required by JWS spec? Can we add a comment here.

Reply:
`resp.Signature` type is `[]byte`, so it is not base64-encoded. The spec says that signature in the response json should be base64-encoded, and `json.[Un]Marshal` takes care of that, but the in-memory representation uses a raw `[]byte` for convencience.